### PR TITLE
[Feature] Add minigpt4 gradio demo and training script

### PR DESCRIPTION
**Pull request by hmtbgc**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Implement gradio demo and training script for MiniGPT-4. 

## Modification

- Implement ``minigpt4_demo.py`` and another auxiliary file (``conversation.py``) under projects/gradio_demo.
- Modify ``tools/train.py`` to support DeepSpeed training.
- Provide dataset for training (``mmpretrain/datasets/minigpt4_dataset.py``).
- Modify ``minigpt4.py`` to train properly.
- Provide a new script (``minigpt-4_baichuan-7b_caption.py``) for training minigpt4. This script replaces vicuna with other LLM to improve Chinese-generating ability.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

No

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

### Training guide 
- Download dataset from [this url](https://huggingface.co/datasets/deepHug/minigpt4_training_for_MMPretrain). There are two datasets: minigpt4 and llava. Both could be used individually for training.  Modify line 37 of ``configs/minigpt4/minigpt-4_baichuan-7b_caption.py`` as path to minigpt4/llava dataset, and line 38 as ``minigpt4_official_data.json``/``llava_official_data.json``.
- Download baichuan7b weight from [this url](https://huggingface.co/baichuan-inc/Baichuan-7B) and modify line 104 and 108 as path to baichuan7b weight.
- Run the following command in terminal:
    ```shell
    export CONFIG_FILE=configs/minigpt4/minigpt-4_baichuan-7b_caption.py
    bash ./tools/dist_train.sh ${CONFIG_FILE} ${NUM_GPUS}
    ```
    Each GPU may take up 28GB memory. It take ~34mins to train 6 epochs. 4 GPUs are used for training by default. If you only have one GPU, the learning rate should be reduced appropriately (e.g. from 1e-3 to 5e-4).
- After training, you will get a file named ``mp_rank_00_model_states.pt`` under ``work_dirs/minigpt-4_baichuan-7b_caption/epoch_6.pth``. Post process `` mp_rank_00_model_states.pt``with the following python code:

    ```python
    import torch
    from collections import OrderedDict
    file_path = '/path/to/mp_rank_00_model_states.pt'

    data = torch.load(filename)['module']

    destination = OrderedDict()
    destination['llama_proj.weight'] = data['llama_proj.weight']
    destination['llama_proj.bias'] = data['llama_proj.bias']
    torch.save(destination, '/path/to/save/model.pth')
    ```

### Testing guide
- Run the following command in terminal:
  ```shell
  python projects/gradio_demo/minigpt4_demo.py /path/to/minigpt-4_baichuan-7b_caption.py /path/to/save/model.pth
  ```
  It needs 20GB GPU memory to start a chat with minigpt4.

### Examples
- **BaiChuan7B + minigpt4_official_data**
en_2
zh_2
en_3
zh_3
- **BaiChuan7B  + llava_official_data**
en_2
zh_2
en_3
zh_3
- **Original Vicuna7B**
en_2
zh_2
en_3
zh_3

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @hmtbgc,

We'd like to express our appreciation for your valuable contributions to the mmpretrain. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: projects/gradio_demo/conversation.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import torch
+
+
+@dataclasses.dataclass
+class Conversation:
+    system: str
+    roles: List[str]
+    messages: List[List[str]]
+    sep: str = '###'
```
Comment (mzr1996) [inline projects/gradio_demo/conversation.py:12]: Is the `conversation` class copied from another repository? If it is, please note at the docstring or the head of the file.

File: tools/train.py [hunk 2]
```diff
@@ -152,7 +153,10 @@ def main():
     # build the runner from config
     if 'runner_type' not in cfg:
         # build the default runner
-        runner = Runner.from_cfg(cfg)
+        if 'strategy' in cfg:
+            runner = FlexibleRunner.from_cfg(cfg)
+        else:
+            runner = Runner.from_cfg(cfg)
```
Comment (mzr1996) [inline tools/train.py:159]: Please use `runner_type` in the config file to specify whether to use `FlexibleRunner`, and don't modify the `train.py` if not necessary.

File: configs/minigpt4/minigpt-4_vicuna-7b_caption.py [hunk 3]
```diff
@@ -57,10 +57,18 @@
...
+            '<Img><ImageHere></Img> Take a look at this image and describe what you notice.',  # noqa
+            '<Img><ImageHere></Img> Please provide a detailed description of the picture.',  # noqa
+            '<Img><ImageHere></Img> Could you describe the contents of this image for me?'  # noqa
+        ],
+        [
+            '<Img><ImageHere></Img> 详细描述这张图片。',
+            '<Img><ImageHere></Img> 浏览这张图片并描述你注意到什么。',
+            '<Img><ImageHere></Img> 请对这张图片进行详细的描述。',
+            '<Img><ImageHere></Img> 你能为我描述这张图片的内容吗？'
+        ]
```
Comment (mzr1996) [inline configs/minigpt4/minigpt-4_vicuna-7b_caption.py:71]: It's better to use a dict from language to prompts as the multi-language prompts. The list requires an order of languages implicitly, which may be confusing.

File: configs/minigpt4/minigpt-4_vicuna-7b_caption.py [hunk 4]
```diff
@@ -57,10 +57,18 @@
     task='caption',
     prompt_template='###Human: {} ###Assistant: ',
```
Comment (mzr1996) [inline configs/minigpt4/minigpt-4_vicuna-7b_caption.py:58]: I noticed that you have replaced the `prompt_template` with `en_prompt_tempalte` and `zh_prompt_template`, and here haven't been modified.
I think you can also use a dict instead of several arguments to specify multi-language prompt templates, which can be more extensible.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: tools/train.py (hunk 2 above)
Comment (hmtbgc) [reply tools/train.py]: ``train.py`` has been restored.

File: projects/gradio_demo/minigpt4_demo.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import numpy as np
+import torch
+from conversation import EN_CONV_VISION, ZH_CONV_VISION, Chat
+
+from mmpretrain import ImageCaptionInferencer
+
+parser = argparse.ArgumentParser(description='MiniGPT4 demo')
+parser.add_argument(
+    '-mp',
+    '--model_path',
```
[the remark this answers is not in the document]
Comment (hmtbgc) [reply projects/gradio_demo/minigpt4_demo.py]: ``cfg`` and ``ckpt`` have been used to replace ``--model_path`` and ``--pretrained_path``

File: projects/gradio_demo/conversation.py (hunk 1 above)
Comment (hmtbgc) [reply projects/gradio_demo/conversation.py:14]: Have added the original repository at the head of ``conversation.py``.

File: configs/minigpt4/minigpt-4_vicuna-7b_caption.py (hunk 3 above)
Comment (hmtbgc) [reply configs/minigpt4/minigpt-4_vicuna-7b_caption.py]: Have changed the type of ``raw_prompt`` and ``prompt_template`` as ``dict``, which has two keys: ``en`` and ``zh``.

File: configs/minigpt4/minigpt-4_vicuna-7b_caption.py (hunk 4 above)
Comment (hmtbgc) [reply configs/minigpt4/minigpt-4_vicuna-7b_caption.py]: Fixed in the above issue.

**Review by mzr1996 — APPROVED** [review]
